qutip/qip/circuit_latex.py

```diff
@@ -42,31 +47,121 @@
+
+
+def _convert_pdf(file_stem):
     """
-    Requires: pdflatex, pdfcrop, pdf2svg, imagemagick (convert)
+    'Convert' to pdf: since LaTeX outputs a PDF file, there's nothing to do.
     """
-    os.system("rm -f %s.tex %s.pdf %s.png" % (filename, filename, filename))
+    with open(file_stem + ".pdf", "rb") as file:
+        return file.read()
+
 
-    with open(filename + ".tex", "w") as file:
-        file.write(_latex_template % (_qcircuit_latex_min, code))
+# Record type to hold definitions of possible conversions - this is just for
+# reading convenience.
+_PossibleConverter = collections.namedtuple(
+    '_PossibleConverter',
+    ['file_type', 'dependency', 'executables', 'arguments', 'binary'],
+)
+CONVERTERS = {"pdf": _convert_pdf}
+_POSSIBLE_CONVERTERS = [
+    _PossibleConverter('png', 'ImageMagick', ['magick', 'convert'],
+                       arguments=('-density', '100'), binary=True),
+    _PossibleConverter('svg', 'pdf2svg', ['pdf2svg'],
+                       arguments=(), binary=False),
+]
 
-    os.system("pdflatex -interaction batchmode %s.tex" % filename)
-    os.system("rm -f %s.aux %s.log" % (filename, filename))
-    os.system("pdfcrop %s.pdf %s-tmp.pdf" % (filename, filename))
-    os.system("mv %s-tmp.pdf %s.pdf" % (filename, filename))
 
-    if format == 'png':
-        os.system("convert -density %s %s.pdf %s.png" % (100, filename,
-                                                         filename))
-        with open("%s.png" % filename, "rb") as f:
-            result = f.read()
+def _make_converter(possible):
+    """
+    Create the actual conversion function of signature
+        file_stem: str -> 'T,
+    where 'T is data in the format to be converted to.
+    """
+    which = _find_system_command(possible.executables)
+    if which is None:
+        return None
+    mode = "rb" if possible.binary else "r"
+
+    def converter(file_stem):
+        in_file = file_stem + ".pdf"
+        out_file = file_stem + "." + possible.file_type
+        _run_command((which, *possible.arguments, in_file, out_file))
+        with open(out_file, mode) as file:
+            return file.read()
+    return converter
+
+
+for possible in _POSSIBLE_CONVERTERS:
+    # Make the converter using a higher-order function, because if we defined a
+    # function in the loop, it would be easy to later introduce bugs due to
+    # leaky closures over loop variables.
+    converter = _make_converter(possible)
+    if converter:
+        CONVERTERS[possible.file_type] = converter
     else:
-        os.system("pdf2svg %s.pdf %s.svg" % (filename, filename))
-        with open("%s.svg" % filename) as f:
-            result = f.read()
+        message = "".join(["Could not find system ", possible.dependency, ".",
+                           " Image conversion to '", possible.file_type, "'",
+                           " is unavailable."])
+        # ImportWarning is ignored by default, so the user won't actually see
```

Comment:
Actually this part of code will be executed even if the user does not intend to plot anything, for example it executes when `from qutip.qip.circuit import QubitCircuit` runs. Is Showing the warning here confusing?

Reply:
This was one of the reasons I put the PR up in the first place - I didn't think this was the best idea, but at the time it was what I thought of.

The reason it is here rather than elsewhere is so that the check for the converters only happens once, and the warning is emitter by the checkers.  It would be very convoluted logic to have the checks put the warnings somewhere, then have them later be emitted when the code is actually accessed.

The warning actually won't show at all to most people - `ImportWarning` is by default ignored by the filters (which is another problem in its own right...).

I wanted to have a conversation about how would be best to implement these warnings, since this clearly isn't perfect.

Reply:
The other reason that I put the warning in at import time is because if the dependency is missing, then the `svg` and `png` methods of `QubitCircuit` aren't added (so IPython doesn't try and introspect them), and so it's not actually really possible to call the function without them.  The user would then never even know that they're missing functionality.

If there's a way to get IPython to _not_ use an existing `class.svg()` method, then we can emit the warning in `make_image` relatively neatly.  We'd add a sub-branch of the `if file_type not in _CONVERTERS` - we add a second check `if file_type in _MISSING_CONVERTERS` to determine whether it's one we could do if we have the dependency, or one that's clearly something totally screwed up in the code.